### PR TITLE
feat: memtable filter push down

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR enables filter pushdown for memtable so that irrelavant time series can be skipped.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Comment by killme2008** [comment]
I think we should add some benchmark for this feature.

**Comment by waynexia** [comment]
> I think we should add some benchmark for this feature.

I've run an informal test against this branch, it can improve by around 70% performance in scanning memtable.

**Comment by killme2008** [comment]
> > I think we should add some benchmark for this feature.
> 
> I've run an informal test against this branch, it can improve by around 70% performance in scanning memtable.

Cool. It would be beneficial for future improvements if we include a benchmark program in the repository.

**Review by waynexia — APPROVED** [review]
LGTM

File: src/table/src/predicate.rs [hunk 1]
```diff
@@ -119,6 +121,39 @@ impl Predicate {
...
+                    predicate_array
+                        .into_iter()
+                        .map(|x| x.unwrap_or(true))
+                        .next()
+                        .unwrap_or(true)
+                }
+                // result was a column
+                ColumnarValue::Scalar(ScalarValue::Boolean(v)) => v.unwrap_or(true),
+                _ => {
+                    unreachable!()
```
Comment (evenyag) [inline src/table/src/predicate.rs:143]: Maybe print `eva` in unreachable?

File: src/mito2/src/memtable/time_series.rs [hunk 2]
```diff
@@ -285,45 +294,137 @@ impl SeriesSet {
...
+
+    let result = predicate.prune_primary_key(&pk_record_batch);
+    debug!(
+        "Prune primary key: {:?}, res: {:?}",
+        pk_record_batch, result
+    );
+    result.unwrap_or(true)
+}
+
+fn pk_to_record_batch(
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/memtable/time_series.rs:472]: > nitpick(non-blocking): Maybe we can cache the decoded record batch in each series. We need a benchmark to see whether it is valuable.

Fixed in https://github.com/GreptimeTeam/greptimedb/pull/2539/commits/441bbc711de9330ea636b0af2b5161068a4782a8#diff-e0e58a966eb1213f9cf956c8e72709d59c9ee606f8aafbcb138fe9740ab88746R458-R460

**Review by evenyag — APPROVED** [review]

**Pull request merged**